llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -883,6 +881,110 @@ static bool removeDeviceGlobalFromCompilerUsed(Module &M) {
   return true;
 }
 
+// Add extra red zone to each image scope device globals if the module has been
+// instrumented by sanitizer pass. And record their infomation like size, red
+// zone size, beginning address.
+static bool instrumentDeviceGlobal(Module &M) {
+  auto &DL = M.getDataLayout();
+  IRBuilder<> IRB(M.getContext());
+  SmallVector<GlobalVariable *, 8> GlobalsToRemove;
+  SmallVector<GlobalVariable *, 8> NewDeviceGlobals;
+  SmallVector<Constant *, 8> DeviceGlobalMetadata;
+
+  constexpr uint64_t MaxRZ = 1 << 18;
+  const uint64_t MinRZ = 32;
+
+  Type *IntptrTy =
+      Type::getIntNTy(M.getContext(), M.getDataLayout().getPointerSizeInBits());
+
+  // Device global meta data is described by a structure
+  //  size_t device_global_size
+  //  size_t device_global_size_with_red_zone
+  //  size_t beginning address of the device global
+  StructType *StructTy = StructType::get(IntptrTy, IntptrTy, IntptrTy);
+
+  for (auto &G : M.globals()) {
+    // Non image scope device globals are implemented by device USM, and the
+    // out-of-bounds check for them will be done by sanitizer USM part. So we
+    // exclude them here.
+    if (isDeviceGlobalVariable(G) && hasDeviceImageScopeProperty(G)) {
```

Comment:
This should be turned into an early exit instead

Reply:
Done